File: framework/lib/controller.dart
part of uxml;

/**
 * Provides base class for controller for MVC framework.
 *
 * @author:ferhat@ (Ferhat Buyukkokten)
 */
class Controller extends UxmlElement {
  static ElementDef controllerElementDef = null;
  /** Controller Property Definition */
  static PropertyDefinition controllerProperty;
  EventHandler _viewInitHandler = null;
  EventHandler _viewClosedHandler = null;

  /**
   * Holds reference to view.
   */
  UIElement _attachedView;

  /**
   * Holds bindings for this elements properties.
   */
  List<Object> bindings;

  Controller(UIElement view) : super() {
    bindings = [];
    _attachedView = view;
    view.setProperty(Controller.controllerProperty, this);
    _viewClosedHandler = onViewClosed;
    view.addListener(UIElement.closedEvent, _viewClosedHandler);
    _viewInitHandler = onViewInitialized;
    view.addListener(Controller.controllerProperty, _viewInitHandler);
  }

  /**
   * Returns attached view.
   */
  UIElement get view => _attachedView;

  /**
   * Returns controller of element.
   */
  static getTargetController(UIElement element) {
    return element.getProperty(Controller.controllerProperty);
  }

  /**
   * Sets controller of element.
   */
  static setTargetController(UIElement element, Controller controller) {
    element.setProperty(Controller.controllerProperty, controller);
  }

  /**
   * Called on chrome controllers before the element tree is assigned.
   */
  void preInit() {
  }

  /**
   * Called by view when initialization is complete.
   */
  void initCompleted() {
  }

  /**
   * Called when view's surface is initialized.
   */
  void viewInitialized() {
  }

  /**
   * Called when view is closed. Override this to perform cleanup.
   */
  void close() {
    if (_attachedView != null) {
      _attachedView.clearProperty(Controller.controllerProperty);
    }
    if (_viewInitHandler != null) {
      view.removeListener(Controller.controllerProperty, _viewInitHandler);
      _viewInitHandler = null;
    }
  }

  /**
   * Responds to view close event.
   */
  void onViewClosed(EventArgs event) {
    _attachedView.removeListener(UIElement.closedEvent, _viewClosedHandler);
    close();
  }

  /**
   * Responds to view surface initialized event.
   */
  void onViewInitialized(EventArgs event) {
    if (_viewInitHandler != null) {
      view.removeListener(Controller.controllerProperty, _viewInitHandler);
      _viewInitHandler = null;
    }
    viewInitialized();
  }

  /** @see UxmlElement.getDefinition. */
  ElementDef getDefinition() => controllerElementDef;

  /** Registers component. */
  static void registerController() {
    Controller.controllerProperty = ElementRegistry.registerProperty(
        "controller", PropertyType.OBJECT,
        PropertyFlags.ATTACHED | PropertyFlags.INHERIT, null, null);
    controllerElementDef = ElementRegistry.register(
        "Controller",
        UxmlElement.baseElementDef,
        [Controller.controllerProperty], null);
  }
}

File: plugin/com.hello.uxml.tools.ui/src/com/hello/uxml/tools/framework/Controller.java
package com.hello.uxml.tools.framework;

import com.hello.uxml.tools.framework.events.EventArgs;
import com.hello.uxml.tools.framework.events.EventHandler;
import com.hello.uxml.tools.framework.events.EventNotifier;

import java.util.EnumSet;

/**
 * Provides base class for MVC controllers.
 *
 * @author ferhat@(Ferhat Buyukkokten)
 */
public class Controller extends UxmlElement {

  protected UIElement view;

  /** Controller Property Definition */
  public static PropertyDefinition controllerPropDef = PropertySystem.register("Controller",
      Controller.class, Controller.class, new PropertyData(null, EnumSet.of(PropertyFlags.Attached,
          PropertyFlags.Inherit)));

  /**
   * Constructor.
   */
  public Controller(UIElement view) {
    this.view = view;
    view.setProperty(controllerPropDef, this);
    view.addListener(UIElement.closedEvent, new EventHandler() {
        @Override
        public void handleEvent(EventNotifier targetObject, EventArgs e) {
          Controller controller = Controller.getTargetController((UxmlElement) targetObject);
          controller.close();
        }
    });
  }

  /**
   * Returns view being controlled.
   */
  public UIElement getView() {
    return view;
  }

  /**
   * Called on chrome controllers before the element tree is assigned.
   */
  public void preInit() {
  }

  /**
   * Provides override that is called after view initialization is complete.
   */
  public void initCompleted() {
  }

  /**
   * Called when view is closed. Override this to perform cleanup.
   */
  protected void close() {
  }

  /**
   * Returns controller of element.
   */
   public static Controller getTargetController(UxmlElement element) {
     return (Controller) element.getProperty(controllerPropDef);
   }

   /**
    * Returns controller of EventNotifier.
    */
    public static Controller getTargetController(EventNotifier element) {
      return (Controller) ((UxmlElement) element).getProperty(controllerPropDef);
    }

   /**
    * Sets controller of element.
    */
   public static void setTargetController(UxmlElement element, Controller controller) {
     element.setProperty(controllerPropDef, controller);
   }
}
